### PR TITLE
Parameter selection

### PR DESCRIPTION
**Pull request by iamjli**
2 updates

 * method to generate robust network summary statistics
 * improved implementation of `get_robust_subgraph_from_randomizations`

### CODE REVIEW

**Comment by zfrenchee** [comment]
Please remove the .DS_Store files.

**Comment by iamjli** [comment]
Done

File: src/graph.py [hunk 1]
```diff
@@ -581,6 +583,46 @@ def grid_search(self, prize_file, Ws, Bs, Gs):
...
+    ###########################################################################
+                #######          Summarize randomization results          #######
+    ###########################################################################
+
+    def generate_basic_statistics(self, robust_results): 
+        """
+        Summarizes robust network randomization results
+
+        Arguments: 
+            robust_results (dict of networkx): networkx robust subgraphs indexed by paramstring
```
Comment (zfrenchee) [inline src/graph.py]: Can you invert this sentence? Maybe just say something like `{paramstring: robust subgraph (networkx)}` to give people a clear picture.

File: src/graph.py [hunk 2]
```diff
@@ -25,6 +25,8 @@
 from networkx.readwrite import json_graph as nx_json
 import community    # pip install python-louvain
 from sklearn.cluster import SpectralClustering
+from scipy import stats
+from scipy.spatial import distance
```
Comment (zfrenchee) [inline src/graph.py]: Seems like you aren't using these in the code anymore. Remove them?

File: src/graph.py [hunk 3]
```diff
@@ -581,6 +583,46 @@ def grid_search(self, prize_file, Ws, Bs, Gs):
...
+                    "mean_robustness":  robust_df.robustness.mean(), 
+                    "max_specificity":  robust_df.specificity.max(),
+                    "mean_specificity": robust_df.specificity.mean(), 
+                    "mean_log_degree": np.log2(robust_df.degree).mean(), 
+                    "std_log_degree": np.log2(robust_df.degree).std(), 
+                    "KS_statistic": stats.ks_2samp(np.log2(robust_df.degree), np.log2(self.node_attributes.degree))[0]
+                }
+                
+        robust_summary = pd.DataFrame.from_dict(robust_summary, orient='index')
+        robust_summary["euclidean_distance"] = ((1 - robust_summary.mean_robustness) ** 2 + robust_summary.mean_specificity ** 2) ** 0.5
```
Comment (zfrenchee) [inline src/graph.py]: Leave a comment here explaining why you're doing this outside of the `for` loop. Also mention what the distance is "from" / "to" which is implicit in this line I think.

File: src/graph.py [hunk 4]
```diff
@@ -716,14 +758,22 @@ def get_robust_subgraph_from_randomizations(nxgraph, max_size=400, min_component
...
+    # not more than `max_size`. This is accomplished by progressively loosening the threshold for inclusion in the subgraph. 
+    for top_n in np.arange(min(max_size, len(node_attributes_df)), len(node_attributes_df)+1): 
+        # Get robust subnetwork and remove small components.
+        robust_network = nxgraph.subgraph(node_attributes_df.index[:top_n])
+        robust_network = filter_graph_by_component_size(robust_network, min_component_size)
+
+        if robust_network.number_of_nodes() <= max_size: 
+            robust_network_out = robust_network
+        else: 
+            return robust_network_out
```
Comment (zfrenchee) [inline src/graph.py]: I'm confused about this else block. What is `robust_network_out` when you go through the else?

File: src/graph.py (hunk 2 above)
Comment (iamjli) [reply src/graph.py]: We're using scipy.stats, removed line 29 though

File: src/graph.py (hunk 3 above)
Comment (iamjli) [reply src/graph.py]: Removed this feature since it's a bit arbitrary.

File: src/graph.py (hunk 4 above)
Comment (iamjli) [reply src/graph.py]: It's an early return statement so the entire for loop does not need to be executed. Once a robust network is found with more nodes than max size, no need to check less stringent thresholds.

File: src/graph.py [hunk 5]
```diff
@@ -25,6 +25,7 @@
 from networkx.readwrite import json_graph as nx_json
 import community    # pip install python-louvain
 from sklearn.cluster import SpectralClustering
+from scipy import stats
```
Comment (zfrenchee) [inline src/graph.py]: If we need scipy, please add that to requirements.txt and setup.py

File: src/graph.py [hunk 6]
```diff
@@ -581,6 +582,45 @@ def grid_search(self, prize_file, Ws, Bs, Gs):
...
+                robust_df = get_networkx_graph_as_dataframe_of_nodes(robust_network)
+                robust_summary[paramstring] = {
+                    "W": paramstring.split("_")[1],
+                    "B": paramstring.split("_")[3],
+                    "G": paramstring.split("_")[5],
+                    "size": len(robust_df), 
+                    "min_robustness":   robust_df.robustness.min(), 
+                    "mean_robustness":  robust_df.robustness.mean(), 
+                    "max_specificity":  robust_df.specificity.max(),
+                    "mean_specificity": robust_df.specificity.mean(),
```
Comment (zfrenchee) [inline src/graph.py]: please un-justify these (remove extra whitespace) or justify the entire dict. I'll leave it up to your preference.

File: src/graph.py (hunk 5 above)
Comment (iamjli) [reply src/graph.py]: Done

File: src/graph.py (hunk 6 above)
Comment (iamjli) [reply src/graph.py]: Done

**Pull request merged**